### ISSUE TITLE
Ticking/Twitching/Creaking due to IO-Support?

### ISSUE THREAD

**Issue by fmessmer** [ros-industrial/universal_robot#106]
Maybe @jrgnicho was right in #92 when he reported a "shaking" of the robot.
For me it is more of a twitching/creaking of the joints, not really constantly shaking with large motion. It's rather very short ticks occuring very infrequently and irregularly (besides the normal humming of the motors).
@jrgnicho: Does this description fit the observations with your hardware?

Could someone third please check and confirm?
As I am out of the office for this week now, I cannot really confirm wether:
- the behavior was there before IO-support was added
- the behaviour is introduced in IO-support set_io service
- the behaviour is introduced in IO-support io_states publishing
- the behaviour is introduced by something else

**Comment by jrgnicho** [comment]
I only observe this twitching behavior when the ur driver is connected but no robot motion is taking place.  This behavior is very frequent, in fact it does occur every time the robot is idled and waiting for a motion command from the ros ur driver.  I couldn't tell with certainty if the IO-support code is the root cause of this issue, I'll revert to a version of the driver that does not have the IO-support in order to see if this still happens.

**Comment by Sticky281** [comment]
Yes, the problem appears when too much data is sent in the state publisher thread of the ur script. It then starts complaining about dropping packages to the joints which causes the twitching/cracking in the joints. In my case this is independent of whether I'm transferring I/O data or other information and appears both during robot motion and when the robot is idle.

**Comment by fmessmer** [comment]
Although, I don't have a solution for the problem, I can give more information on this:
With every "tick" of the robot, there is the following message in the controller log:

```
R  0008d09h17m53.424s  SafetySys  C19A0:  Master had no data to send to joints
R  0008d09h17m53.432s  SafetySys  C29A1:  Ethernet package loss detected from PC to robot
```

(I guess that's the warning/error @Sticky281 is referring to!)

I can also confirm, that this behavior is definitely related to adding the IO capability!
(When I remove/comment all related parts in `prog` and `driver.py` the behavior is fine and the warnings do not show anymore!)

I'm not too familar with how the driver actually works, but looking at the code ([driver.py#L346](https://github.com/ros-industrial/universal_robot/blob/hydro-devel/ur_driver/src/ur_driver/driver.py#L346)) the way `mtype == MSG_GET_IO` is handled, looks different (several `ẁhile` loops) from the others. Also, (to me) it is not clear, how many `bytes` actually are read.

Furthermore, in [prog#L72](https://github.com/ros-industrial/universal_robot/blob/hydro-devel/ur_driver/prog#L72) there are 10 digital_in and 10 digital_out signals (0...9), althought there are only 8 signals available on the controller. Or is this 8+2 for the tool input/output?

I also don't quite understand what the `flags``actually are.

And as a last question:
How is `ToolData` ([deserialize.py#L103](https://github.com/ros-industrial/universal_robot/blob/hydro-devel/ur_driver/src/ur_driver/deserialize.py#L103)) and `MasterBoardData` ([deserialize.py#L118](https://github.com/ros-industrial/universal_robot/blob/hydro-devel/ur_driver/src/ur_driver/deserialize.py#L118)) related to all this?

It seems getting the IO capability fixed needs some serious revision of the current `ur_driver`!

**Comment by Sticky281** [comment]
Yes, those are the errors I was referring to. My guess so far has been that this is related to URScript simply being too slow and hence we've been planning on giving the C API a try instead. 

I just googled the error and according to the <a href="http://ur-update.dk/URsupport/Manuals/ServiceManual/CB2/UR10/ServiceManual_UR10_en_1.02.pdf">UR10 service manual</a> (page 62) the error is caused by "Possible CPU-overload due to structure of user program" and the suggested fix is "Restructure user program", which more or less confirms my assumption that URScript can't handle the additional data transfer (as mentioned previously, the ethernet package loss is independent of sending IO data - it also occurs when sending other additional data).

**Comment by gavanderhoorn** [comment]
> With every "tick" of the robot, there is the following message in the controller log:
> 
> ```
> R  0008d09h17m53.424s  SafetySys  C19A0:  Master had no data to send to joints
> R  0008d09h17m53.432s  SafetySys  C29A1:  Ethernet package loss detected from PC to robot
> ```
> 
> (I guess that's the warning/error @Sticky281 is referring to!)
> 
> I can also confirm, that this behavior is definitely related to adding the IO capability!
> (When I remove/comment all related parts in prog and driver.py the behavior is fine and the warnings do not show anymore!)

This is most likely a performance issue: the many `pow(..)` invocations in the `statePublisherThread()` are probably not helping (it's all just interpreted, so _some_ overhead is expected).

The fact that the _entire_ body of `send_joint_state()` is inside an `enter_critical` / `exit_critical` block is also not too nice.

> I'm not too familar with how the driver actually works, but looking at the code ([driver.py#L346](https://github.com/ros-industrial/universal_robot/blob/hydro-devel/ur_driver/src/ur_driver/driver.py#L346)) the way `mtype == MSG_GET_IO` is handled, looks different (several `while` loops) from the others. Also, (to me) it is not clear, how many `bytes` actually are read.

It's actually quite similar to fi `MSG_WRENCH` deserialisation: the while loops just make sure the buffer contains enough bytes, and if not, to keep reading from the socket (potentially blocking). This is all on the PC side though.

As to the nr of bytes: 9 \* 4 == 36 (9 \* `buf = buf[4:]`). Which is consistent with the nr of IO related `socket_send_int(..)` invocations in `statePublisherThread()`.

> Furthermore, in [prog#L72](https://github.com/ros-industrial/universal_robot/blob/hydro-devel/ur_driver/prog#L72) there are 10 digital_in and 10 digital_out signals (0...9), althought there are only 8 signals available on the controller. Or is this 8+2 for the tool input/output?

I think it is indeed digital+tool. According to the manual there are 8 digital in/out (DO/I 0-7) + 2 tool in/out (DO/I 8-9) (`user_manual_en_UR5_Global.pdf`, section 2.5). Annoyingly, the scripting manual does not provide any information on the valid range of in/out indices. It does for the flags (scriptmanual_en.pdf, v1.8, page 31, `get_flag(n)`): (0, 32) (but that is most likely incorrect, see #108).

> I also don't quite understand what the `flags`actually are.

This is quite common in industrial controllers: it's essentially a global array of 32 booleans, which you can use for inter-programming signalling. An example could be a (poor-mans) mutual exclusion flag, or a debug 'word'.

> How is `ToolData` ([deserialize.py#L103](https://github.com/ros-industrial/universal_robot/blob/hydro-devel/ur_driver/src/ur_driver/deserialize.py#L103)) and `MasterBoardData` ([deserialize.py#L118](https://github.com/ros-industrial/universal_robot/blob/hydro-devel/ur_driver/src/ur_driver/deserialize.py#L118)) related to all this?

At the moment they aren't, but it seems like they should. It's probably way more efficient to just deserialise the `ToolData` and `MasterBoardData` parts from the state info broadcasted by the controller anyway, than trying to get the poor URScript interpreter to bitbang a large nr of booleans as fast as it can (afaict: there isn't any `sleep` in there).

That would require some changes to the Python side though: it currently only uses the `RobotState` data if it is not in `EXECUTING` state.

> It seems getting the IO capability fixed needs some serious revision of the current `ur_driver`!

Your suggestion to use the `ToolData` and `MasterBoardData` is a good one: we can essentially do away with the IO related parts in `statePublisherThread()`. The bits handling writes to IO elements should not be problematic. We would loose the temporal consistency (ie: snapshot behaviour) of the current implementation though.

**Comment by fmessmer** [comment]
I played around a bit, trying to restructure `prog` and `driver.py` a bit.
I tried to fix this with only minor changes, such as splitting up the big thread `statePublisherThread()` into three separate threads for joint_states, wrench and io_states respectively or resolving the `pow()`'s keeping up the current structure.
I did this in dedicated [branch](https://github.com/ipa-fxm/universal_robot/tree/issue106_temporary_fix_restructure_prog).
I think this is a branch that should not be sent as a PR as it also includes several other PRs, i.e. #102, #103, #110, #111, #114, #115, #116 
Maybe we should just use it for discussion and testing. For the final solution a new PR can be sent.

In the branch, I also had to removed sending the flag_states as well as disabling the "flag support" in the `io_interface` helper and the `test_io.py` as the restructuring improved the behavior but did not solve it.
Thus, I guess we definitely need to investigate making use of `ToolData` and `MasterBoardData` to finally solve it and still provide all functionality.

Any thoughts on this?

**Comment by gavanderhoorn** [comment]
If this is really a performance issue (as in: it is not just 'us' being stupid in some way), then we could look into making use of some more of the data in `RobotState`. We should investigate whether all data we need is actually present in that binary stream, then update the driver to make use of it.

Something I'd like to see answered before we decide to reimplement things: what was the rationale for implementing the driver the way it is now? Obviously someone made the effort of implementing `deserialize.py`, but `driver.py` prefers to not use it.

Based on my (naive) understanding of how things work, I'd have made `driver.py` _always_ use the data in `RobotState`, and only have `prog` handle incoming msgs (ie: perform motion based on waypoints, etc). This would seem to incur the least amount of overhead, while still allowing access to the important states.

**Comment by fmessmer** [comment]
Alright, I made an proof-of-concept attempt in order to see how things would work when we decide to use `RobotState` to send IO information (and probably other kind of data).  
You can find the result on this [branch](https://github.com/ipa-fxm/universal_robot/tree/issue106_io_states_from_robot_state_data). Again, I simply started coding, thus, in order to send a PR it needs to be cleaned up.

In that branch, I used the `MasterboardData` structure from `RobotState` to fill the IOStates message. However, as e.g. the `analog_input2` and `analog_input3` are related to the tool they are not within `MasterboardData` but within `ToolData`. So I decided to not have those read from the robot state for less confusion. Also, I could not find anything `Flag`-related in the `RobotState`. So this is still send from `prog`.  
As you can be seen, it is quite simple to fill the messages in the `__on_packet` function.

This function is called at about 10Hz. Which is probably the reason for having `prog` send the `joint_states` (at 125Hz). But 10Hz would definitely be enough for the IO-States, I guess.

One more thing I noticed:
The value send for analog IO's depend on additional parameters such as range and thus cannot be interpreted by simply multiplying them with `MULT_analog`.

And then one more suggestion:
In case we decide to send IO info (and others) using the `RobotState`, I'd also suggest to define new message types such as `MasterboardData.msg` to better reflect the actual structure. Then it is easy to actually publish all the data available (e.g. including the range info for analog values).

We then still can discuss whether:
- we keep `joint_states` being published as is (I vote for this!)
- we keep including the flags or remove them (I vote for removing as flags are not used within `prog` which is the script being executed on the ur controller)
- we keep setting the IO states through `prog` (I guess this is the only way possible)

Any comments?

**Comment by gavanderhoorn** [comment]
Have you considered using the _real-time client-interface_? Should give you about the same data as `RobotState` now, but at 125Hz. You cannot use `RobotState` as it is right now though. The packet layout is different.

**Comment by fmessmer** [comment]
I don't know about this interface. Can you send me a Link or some reference?

---

EDIT: 
Found it on the wiki (http://wiki03.lynero.net/Technical/RealTimeClientInterface?rev=9)

I guess, I could give it a shot....at least proof-of-concept.
Once we see, it works that way we should open up a fresh issue where we discuss what the **NEW** ur_driver is supposed to look like, i.e. which information from which port mapped to which message type published to which topic...

**Comment by gavanderhoorn** [comment]
@ipa-fxm: that is the one. But I'm not sure whether we are 'allowed' to implement a deserialiser for those packets in this driver. Also, I've had trouble using that documentation in the past. All field types are listed as `double` fi, but I'm pretty sure there were some `integer`s in there as well.

As a _new_ driver: I'm guessing the _ros->urscript_ bits can be almost left as is. It's just the state reporting that seems to cause performance issues.

**Comment by fmessmer** [comment]
FYI, in this [branch](https://github.com/ipa-fxm/universal_robot/tree/deserialize_realtime_communicationsinterface) I added the deserializer for the data structure that is send through the real-time communication interface (Socket 30003).
Additionally, I added a new ur_mgs.msg `RobotStateRTMsg` reflecting that struture.

The communication can be tested using the `ur_driver/testRT_comm.py` script.
I also added a new connection (`URConnectionRT`) to the driver. It's an addition! I haven't removed anything from the current driver (yet).

While performance is fine using the test script, when included in the driver the buffer for the real-time socket connection fills up as the packages cannot be processed quickly enough (thus, `/joint_states` starts lagging behind). This probably can be solved by removing duplicate parts of the driver that now will be handled by the `URConnectionRT` (e.g. the parsing of the `prog` protocol is still active in the branch).

The values received and published on `/robot_stateRT` all seem plausible, but more testing is needed!

**Comment by gavanderhoorn** [comment]
@ipa-fxm: deserialisation performance of `struct.unpack(..)` should be enough for this: the deserialising routine can probably be made a bit more efficient. Also, I doubt whether the other parts of the (Python) driver are an issue here: they should have their own connection buffer.

While making more use of the `RobotState` data (and its RT counterpart) could solve some of the issues we encountered, there are some disadvantages to it:
1. UR seems to change the pkg layout from time to time, which would require the driver maintainer(s) to keep the deserialiser up-to-date. The current URScript `prog` does not suffer from this (the Python bits do, but only in a minor way, see #9 and #16).
2. Not all state we're (potentially) interested in is broadcast on TCP30003 (analog in/outputs fi). Unlike the URScript driver, we cannot change the contents of the RT-state packet.
3. An increase in resource use: the processing overhead can probably be reduced, but is an issue none-the-less.

The major advantage is the -- essentially free -- additional robot state information we get, while avoiding overloading the URScript interpreter.

PS: I assume the `RobotStateRTMsg` is for debugging purposes only? For an official driver, I'd stick to `JointStates` and other appropriate msgs.

**Comment by fmessmer** [comment]
PR #117 should fix this issue.
Please review!

**Comment by gavanderhoorn** [comment]
Just to document this a bit more, here are some results from a session with `sar` (from `sysstat`) on our UR5 (v1.8.14035).

I installed [sysstat](http://sebastien.godard.pagesperso-orange.fr/index.html) on the controller (using the debian lenny [package](http://archive.debian.net/eo/lenny/sysstat)) and tried to monitor resource usage on the controller.

This is all with the 'normal' version of `prog` ([ba8e2270](https://github.com/ros-industrial/universal_robot/blob/ba8e2270f2d9af32fc47077fa88deade7dd9fcd9/ur_driver/prog), `hydro-devel`).

---

A completely idle controller (no driver loaded, no motion, initialised, robot on), shows ~ 4-5 % CPU utilisation on both cores.

Running `prog` (no motion, initialised, robot on), I get a cpu utilisation of ~ 20% on core 0, ~ 7-8% on core 1 (core 0 probably runs `URControl`).

Looking at network traffic, an idle controller has about 23 / 11 KB/s (rx, tx) on `eth0` (which I assume is connected to the motion controller hw), no traffic on `eth1` (expected, no one is connected).

Network traffic for controller with driver: `eth0`: 23 / 11 KB/s, `eth1`: 43 / 65 KB/s (rx, tx). `eth1` is now used to communicate with the PC.

Interrupt rate in both cases shows a delta of about 1500 intr/s: ~3500 intr/s idle, ~5000 intr/s with driver loaded.

---

This was all with the Polyscope application showing just the main screen.

Looking at the _log_ tab of the _PROGRAM robot_ interface showed no warning / errors for the idle controller (or at least, nothing not seen for a normal boot), while the controller with the driver loaded showed:

```
...
R __TIMESTAMP__ URControl C116A117: Heavy processor load warning
R __TIMESTAMP__ SafetySys C19A0: Master had no data to send to joints
R __TIMESTAMP__ SafetySys C29A1: Ethernet package loss detected from PC to robot
T __TIMESTAMP__ RTMachine Program driverProg started
T __TIMESTAMP__ RTMachine socket_read_binary_integer: timeout
T __TIMESTAMP__ RTMachine socket_read_binary_integer: timeout
...
```

after which it starts to repeat the following two lines:

```
R __TIMESTAMP__ SafetySys C19A0: Master had no data to send to joints
R __TIMESTAMP__ SafetySys C29A1: Ethernet package loss detected from PC to robot
```

I assume `C29A1` complains about traffic on `eth0`, because Wireshark showed no package loss on the connection over `eth1`.

`C116A117` is only seen once, when the driver (`prog`) is being loaded. That is probably not really an issue.

Perhaps whatever causes `C19A0` is the cause of the "ticking / twitching / creaking" reported in this ticket.

---

What puzzles me is that the data measured do not directly seem to indicate a performance issue. Perhaps the tools used do not have the necessary resolution, or I'm misinterpreting things.

**Comment by gavanderhoorn** [comment]
Repeating the same measurements for the version of `ur_driver` proposed in #117:
- no errors / warnings are reported (apart from a the timeout of `socket_read_binary_integer(..)`)
- CPU usage is about 10% for core 0, 5% for core 1
- Interrupt rate is at ~ 3700 intr/s
- Network traffic is similar to the earlier figures for `eth0`, but `eth1` now shows about 7 / 113 KB/s (rx, tx). This is expected, as the _real-time client-interface_ sends out 812 byte pkts @ 125Hz.

Average packets / second seems to have gone down though:
- current (`hydro-devel`) version of `prog`: ~ 710 pkts/s when connected, both rx and tx
- proposed version: ~ 120 pkts/s when connected, again rx and tx

Perhaps the overhead incurred by the larger nr of packets (coming from the current version of `prog`) on `eth1` is something the controller cannot deal with.

**Comment by abubeck** [comment]
Is resolved with #117 .
